include: util: Add `memeq` and `bool_memcmp`

theob-pro opened the pull request:
Those two functions are used to compare memory area. `bool_memcmp` is basically `memcmp` but return a `bool` instead of an `int`. `memeq` compare the length of the memory area in addition to the bytes themselves.

theob-pro committed:
include: util: Add `util_eq` and `util_memeq`

`util_eq` compare two memory areas and their length. It returns true if
they are equal, else false.

`util_memeq` the first n bytes of two memory areas. It returns true if
they are equal, else false.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Tests: Util: Add unit tests for `util_eq`/`util_memeq`

Add two unit tests for the newly added functions `util_eq` and
`util_memeq`.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

pdgendt commented:
I don't see the benefit of adding these, sorry.

pdgendt commented:
> No worries, it may not be useful for everyone 😀 More context on the discussion I had with @Thalley on this PR: #84018

What I mean was that I don't see any added value of introducing what basically is an alias...

```c
if (memeq(buf1, size1, buf2, size2)) { /* ... */ }
/* vs */
if (size1 == size2 && memcmp(buf1, buf2, size1) == 0) { /* ... */ }
```

andyross commented:
Tend to agree with @pdgendt, we don't want to be pushing APIs that compete with well-established controlling idioms (like the ISO C standard memcmp() in this case, but we take stuff from Linux too).

theob-pro commented:
> What I mean was that I don't see any added value of introducing what basically is an alias...

Maybe I am missing something but that file is full of aliases... I don't see the problem with adding new ones

> Tend to agree with @pdgendt, we don't want to be pushing APIs that compete with well-established controlling idioms (like the ISO C standard memcmp() in this case, but we take stuff from Linux too).

I struggle to understand the reasoning about APIs that compete, sorry, could you develop? And those function are just more convenient to use. Nobody is forced to use them 😅

theob-pro commented:
> I disagree. The file, in my opinion, provides common functions people tend to get wrong or inefficient implementations of.

Sorry but I must insist, there is an alias for `is_power_of_two`, macro for `IN_RANGE`, `MIN` and `MAX`, a macro for casting a pointer to an unsigned integer. Those functions and macro are there for quality of life, for making code more readable in the end.

> Nobody is forced to use them, but suddenly `memeq` is almost globally defined as `zephyr/sys/util.h` will likely be included in some way or another.

Sorry, I don't understand, what is the problem here? It will not be included in people code if they don't use it

Thalley commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,46 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+ * @returns true if both the length of the memory areas and their content are
+ * equal else false
+ */
+static inline bool memeq(const void *m1, size_t len1, const void *m2, size_t len2)
+{
+	int ret;
+
+	if (len1 != len2) {
+		return false;
+	}
+
+	ret = memcmp(m1, m2, len1);
+
+	return ret == 0;
+}
```
```suggestion
static inline bool memeq(const void *m1, size_t len1, const void *m2, size_t len2)
{
	return len1 == len2 && bool_memcmp(m1, m2, len1);
}
```

Suggest a less verbose version of this function

Thalley commented:
> > I disagree. The file, in my opinion, provides common functions people tend to get wrong or inefficient implementations of.
> 
> Sorry but I must insist, there is an alias for `is_power_of_two`, macro for `IN_RANGE`, `MIN` and `MAX`, a macro for casting a pointer to an unsigned integer. Those functions and macro are there for quality of life, for making code more readable in the end.
> 
> > Nobody is forced to use them, but suddenly `memeq` is almost globally defined as `zephyr/sys/util.h` will likely be included in some way or another.
> 
> Sorry, I don't understand, what is the problem here? It will not be included in people code if they don't use it

I agree with @theob-pro here. 

> What I mean was that I don't see any added value of introducing what basically is an alias...

If you apply that argument to the existing `util.h` or `util_macro.h`, then basically none of the macros or functions should be there. Their whole point is being a semantic alias for common operations that makes it easier for the reader to read the code. Even the macro you added yourself (https://github.com/zephyrproject-rtos/zephyr/pull/66744) wouldn't have been added if you applied that argument.

Thalley commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,46 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+ *
+ * @returns true if the @p n first bytes of @p m1 and @p m2 are the same, else false
+ */
+static inline bool bool_memcmp(const void *m1, const void *m2, size_t n)
```
I dislike the naming of this function. 

IMO `memeq` should be `memcmp(m1, m2, n) == 0;` and the function that implements the size check too, that should call this function, should have the longer name. Maybe `memeqlen`? 

We could also apply longer names to avoid clashes or confusion with the standard libraries. 

How about `mem_is_equal` and `mem_is_equal_with_len`?

alwa-nordic replied:
I think it should be the other way around. I think the default notion of equality should include the length check, just like it is defined in "most" programming languages (usually called `==`). That is: It determines two memory slices to be equal if and only if both their lengths and contents are the same. It determines two memory slices of different lengths as different always, and two slices of zero length as equal always. This is true for C++ (vector ==), Python (bytearray ==), Go (slices.Equal), Rust (Slice ==) and pretty much everything else.

Looking at the man page for `memcmp`, I conclude that `memcmp` is defined as taking two memory slices that are known to be of equal length. So as an *optimization*, it takes only one length argument. (The result of `memcmp` is the ordering of the two input slices.) So, I argue, `memcmp` taking one length argument is the special case. The programmer needs to make sure the slices are of equal lengths.

alwa-nordic replied:
To illustrate what I mean by saying `memcmp` taking one length argument is an optimization: Extending the notion of ordering for memory, so that it matches `strcmp`, the shorter memory slice should be ordered first. So `full_memcmp(s1, l1, s2, l2) := l2 - l1 || memcmp(s1, s2, l1)`.

Thalley replied:
> Looking at the man page for `memcmp`, I conclude that `memcmp` is defined as taking two memory slices that are known to be of equal length.

The manpage of `memcmp` specifically states that it compares the first `n` octets; `n` here is not a reference to either of the inputs' sizes. This is also the reason why I think a single argument is better since it effectively calls `memcmp`, as it should only compare the first `n` octets where `n` may be equal to or smaller than either of the sizes. 

For the Python, C++, etc. reference those are reference to objects which are semantically different than "just memory". 

As pointed out earlier, the default case is comparing memory of equal size, as evident by the Zephyr code base. If 9/10 times are only comparing the first `n` octets, and not the sizes, then that should be the default case. Comparing the size before calling memcmp is the atypical case. 

If we were discussing a specific implementation, e.g. a `net_buf` where we would have a `net_buf_equal(const struct net_buf *a, const struct net_buf *b)` then I agree that it would make sense to check the length fields first by default. 

In any case, this is not a hill I'll die on, so I'm not going to "fight" for one more than the other, but I do think that the simpler and more common case should be the "default". 

We could also do something like `mem_eq` with just a single length field and then `obj_eq` that takes the size of each "object".

pdgendt commented:
> > Sorry, I don't understand, what is the problem here? It will not be included in people code if they don't use it
> 
> I agree with @theob-pro here.

This maybe is less of an issue, but still valid to point out that if anyone uses `memeq` as a name in downstream code, this will now trigger a compiler warning/error if `-Wshadow` is active.

> > What I mean was that I don't see any added value of introducing what basically is an alias...
> 
> If you apply that argument to the existing `util.h` or `util_macro.h`, then basically none of the macros or functions should be there. Their whole point is being a semantic alias for common operations that makes it easier for the reader to read the code. Even the macro you added yourself (#66744) wouldn't have been added if you applied that argument.

Fair point, but there is a readability difference between the following IMO, especially if wrapped in some bigger macro declaration.

```c
IF_DISABLED(CONFIG_SOME_OPTION, (bool my_option;))
/* vs */
COND_CODE_1(CONFIG_SOME_OPTION, (), (bool my_option;))
```

Thalley commented:
> This maybe is less of an issue, but still valid to point out that if anyone uses `memeq` as a name in downstream code, this will now trigger a compiler warning/error if `-Wshadow` is active.

I agree. I also suggest some naming that is less similar to stdlib in https://github.com/zephyrproject-rtos/zephyr/pull/84159#discussion_r1922156218. I think it's risky to name is `memeq` even if that is a good name (which is why it is also risky, as that name has been suggested to be added to stdlib and/or specific compilers over the years). 

IMO all Zephyr-specific util functions should have a unique prefix that avoid clashes with other commonly named util functions/macros, e.g. with `z_memeq` or `util_memeq`, but that's an entirely different discussion :D

Thalley reviewed (dismissed):
It would be nice to include at least one use of these functions, i.e. replace one of the `memcmp(..) == 0` cases with this as an illustration of the purposes of them

nashif commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,40 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+ * @returns true if the @p n first bytes of @p m1 and @p m2 are the same, else
+ * false
+ */
+static inline bool z_util_memeq(const void *m1, const void *m2, size_t n)
```
z_ shall not be used for any public APIs, it is a intended to be used for internal/private APIs only.

Thalley replied:
Is that documented anywhere?

The same file has 
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L393-L394
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L731-L745
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util.h#L826-L830

nashif replied:
> Is that documented anywhere?

never made it in the docs, https://github.com/zephyrproject-rtos/zephyr/pull/62198/files. z_ was added long time ago to replace `_` usage in the tree, but unfortuantly it was seen by many as the zephyr designated prefix for APIs and we ended up with public APIs using `z_`....

theob-pro replied:
This is something that must be documented. It seems that there is quite a bunch of public APIs now using the `z_`, is there good reasons to continue to have that hidden rule?

To me it seems that without proper documentation some more will definitely be added and then I don't see the point of blocking it here.

Thalley replied:
@nashif do you have alternative suggestions for naming?

theob-pro replied:
I updated both names

theob-pro committed:
Bluetooth: Add usage of `util_eq` and `util_memeq`

Update `bt_irk_eq` to use `util_memeq` instead of `memcmp` and the
"disconnect" BabbleSim test to use `util_eq` instead of a first
assertion on the size followed by a `memcmp`.

This is done as an example usage of the two new functions.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro commented:
> It would be nice to include at least one use of these functions, i.e. replace one of the `memcmp(..) == 0` cases with this as an illustration of the purposes of them

I added two example usage in the Bluetooth subsystem.

theob-pro commented:
Rebased on main to resolve conflict.

yperess commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,40 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+ */
+static inline bool z_util_eq(const void *m1, size_t len1, const void *m2, size_t len2)
+{
+	return len1 == len2 && z_util_memeq(m1, m2, len1);
```
Nit: I recommend shorting the call to `z_util_memeq` with `m1 == m2 || z_util_memeq(m1, m2, len)`

theob-pro replied:
But then the meaning is different or am I missing something?

theob-pro replied:
I made the change as I was understanding it. Tell me if I was misunderstanding you :slightly_smiling_face:

Thalley reviewed (approved):
Suggest to add an entry to release notes about the new APIs (but not required)

kartben commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,40 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+ * @returns true if the @p n first bytes of @p m1 and @p m2 are the same, else
+ * false
+ */
+static inline bool util_memeq(const void *m1, const void *m2, size_t n)
```
please mention these two new functions in the release notes

theob-pro replied:
Created a following PR to add the functions to the release notes: https://github.com/zephyrproject-rtos/zephyr/pull/88589
